returnn/training.py

```diff
@@ -379,12 +388,78 @@ def run(self):
                             print(f.read())
                     except Exception as exc:
                         print("Cannot read:", exc)
-            sys.stdout.flush()
 
         env = os.environ.copy()
         env["OMP_NUM_THREADS"] = str(self.rqmt["cpu"])
         env["MKL_NUM_THREADS"] = str(self.rqmt["cpu"])
-        sp.check_call(self._get_run_cmd(), env=env)
+
+        rdzv_addr = None
+        if (self.multi_node_slots or 1) > 1 and self.distributed_launch_cmd == "torchrun":
+            import hashlib
+            import psutil
+
+            assert "SLURM_JOB_NODELIST" in os.environ, (
+                "multi-node multi GPU-training w/ torchrun currently depends on SLURM environment variables "
+                "to determine the master node"
+            )
+
+            partaking_nodes = os.environ["SLURM_JOB_NODELIST"]
+            nodes = sorted(
+                node_name.strip()
+                for node_name in sp.check_output(["scontrol", "show", "hostnames", partaking_nodes])
```

Comment:
I don't understand, why is this need? You already have the nodes already (`partaking_nodes`)? How is `nodes` different?

Reply:
`partaking_nodes` is in SLURMs short abbreviated format, e.g. you'll get something like `g-[11-13],g-16`, and that command "decodes" the short form into the full list of proper node names.

Reply:
In any case, I agree to @michelwi (see his comment above), we should not make this logic dependent on Slurm here in the job.